runtime/lua/vim/treesitter/query_linter.lua

```diff
@@ -0,0 +1,146 @@
+local namespace = vim.api.nvim_create_namespace('vim.treesitter.query_linter')
+-- those node names exist for every language
+local MAGIC_NODE_NAMES = { '_', 'ERROR' }
+
+local M = {}
+local lints = {}
+
+local function show_lints(buf, lints)
+  local diagnostics = vim.tbl_map(function(lint)
+    return {
+      lnum = lint.range[1],
+      end_lnum = lint.range[3],
+      col = lint.range[2],
+      end_col = lint.range[4],
+      severity = vim.diagnostic.ERROR,
+      message = lint.message,
+    }
+  end, lints)
+  vim.diagnostic.set(namespace, buf, diagnostics)
+end
+
+local function add_lint_for_node(node, buf, error_type, complete_message)
+  local node_text = vim.treesitter.get_node_text(node, buf):gsub('\n', ' ')
+  local error_text = complete_message or error_type .. ': ' .. node_text
+  local error_range = { node:range() }
+  table.insert(
+    lints[buf],
+    { type = error_type, range = error_range, message = error_text, node_text = node_text }
+  )
+end
+
+local function table_contains(predicate, table)
```

Comment:
vim.tbl_contains ?

Reply:
this file uses both this and `vim.tbl_contains`. This is a dynamic version with a function predicate instead of a value. Semantics is the same (but I should swap parameters)

Reply:
Would it make sense to extend `vim.tbl_contains` to also accept a predicate? (similar to `vim.tbl_filter`)

Reply:
(Note to self: the issue here is that you can't compare tables since `{ foo, bar } == { 'foo', true }` is `false even if `foo == 'foo'` and `bar == true`. This strikes me as a more general problem that could profit from a general solution.)

Reply:
it has wholes in the indices. Try `print(vim.inspect(vim.treesitter.language.inspect("cpp")))`. For me it does not have index 398 but 397 and 399

Reply:
oh, it can mix tables and array-like maps...

Reply:
Maybe add a  `todo` comment for this function along the lines of https://github.com/neovim/neovim/pull/22784#discussion_r1150756969; this can be handled in a followup refactor PR where we can discuss the best approach. No need to hold up this PR with it.

Reply:
Added todo